src/nxt_h1proto.c

```diff
@@ -946,31 +964,72 @@ nxt_h1p_request_body_read(nxt_task_t *task, nxt_http_request_t *r)
 
     r->body = b;
 
-    body_rest = body_length;
+    body_rest = r->chunked ? 1 : body_length;
 
     in = h1p->conn->read;
 
     size = nxt_buf_mem_used_size(&in->mem);
 
     if (size != 0) {
-        size = nxt_min(size, body_length);
-
         if (nxt_buf_is_file(b)) {
-            res = nxt_fd_write(b->file->fd, in->mem.pos, size);
-            if (nxt_slow_path(res < (ssize_t) size)) {
-                status = NXT_HTTP_INTERNAL_SERVER_ERROR;
-                goto error;
-            }
+            if (r->chunked) {
+                out = nxt_http_chunk_parse(task, &h1p->chunked_parse, in);
 
-            b->file_end += size;
+                if (h1p->chunked_parse.error) {
+                    status = NXT_HTTP_INTERNAL_SERVER_ERROR;
+                    goto error;
+                }
+
+                if (h1p->chunked_parse.chunk_error) {
+                    status = NXT_HTTP_BAD_REQUEST;
+                    goto error;
+                }
+
+                for (chunk = out; chunk != NULL; chunk = chunk->next) {
+                    size = nxt_buf_mem_used_size(&chunk->mem);
+
+                    res = nxt_fd_write(b->file->fd, chunk->mem.pos, size);
+                    if (nxt_slow_path(res < (ssize_t) size)) {
+                        status = NXT_HTTP_INTERNAL_SERVER_ERROR;
+                        goto error;
+                    }
+
+                    b->file_end += size;
+
+                    if ((size_t) b->file_end
+                         > r->conf->socket_conf->max_body_size)
```

Comment:
I think the `>` should be aligned to be under the second `(` (i.e pull it back one character) as it's really a continuation of the opening `(`.

Reply:
I've never seen anything like this before, I'm fine with it.
But it reminds me to go the other way. Recheck it, please.

Reply:
I simply meant this

```c
+                    if ((size_t) b->file_end
+                        > r->conf->socket_conf->max_body_size)
```

as if the cast wasn't there...

But you've changed it and it doesn't need splitting over two lines now anyway...